File: src/fr/geobert/radis/data/Operation.java
package fr.geobert.radis.data;

import java.text.ParseException;
import java.util.Calendar;
import java.util.GregorianCalendar;

import android.database.Cursor;
import android.os.Parcel;
import android.os.Parcelable;
import fr.geobert.radis.db.InfoTables;
import fr.geobert.radis.db.OperationTable;
import fr.geobert.radis.tools.Formater;
import fr.geobert.radis.tools.Tools;

public class Operation implements Parcelable {
	protected GregorianCalendar mDate;
	public String mThirdParty;
	public String mTag;
	public String mMode;
	public String mNotes;
	public long mSum;
	public long mScheduledId;
	public long mRowId;
	public String mTransSrcAccName;

	// if these value are != 0, it is a transfert operation between 2 accounts
	// mTransferAccountId is the other account
	public long mTransferAccountId;
	public long mAccountId;

	public Operation(Operation op) {
		mDate = new GregorianCalendar();
		mDate.setTimeInMillis(op.getDate());
		Tools.clearTimeOfCalendar(mDate);
		mThirdParty = op.mThirdParty;
		mTag = op.mTag;
		mMode = op.mMode;
		mNotes = op.mNotes;
		mSum = op.mSum;
		mScheduledId = op.mScheduledId;
		mTransferAccountId = op.mTransferAccountId;
		mTransSrcAccName = op.mTransSrcAccName;
		mAccountId = op.mAccountId;
	}

	public Operation(Cursor op) {
		mThirdParty = op.getString(op
				.getColumnIndexOrThrow(InfoTables.KEY_THIRD_PARTY_NAME));
		if (null == mThirdParty) {
			mThirdParty = "";
		}
		mMode = op.getString(op
				.getColumnIndexOrThrow(InfoTables.KEY_MODE_NAME));
		if (null == mMode) {
			mMode = "";
		}
		mTag = op.getString(op
				.getColumnIndexOrThrow(InfoTables.KEY_TAG_NAME));
		if (null == mTag) {
			mTag = "";
		}
		mSum = op.getLong(op.getColumnIndexOrThrow(OperationTable.KEY_OP_SUM));
		mDate = new GregorianCalendar();
		mDate.setTimeInMillis(op.getLong(op
				.getColumnIndexOrThrow(OperationTable.KEY_OP_DATE)));
		Tools.clearTimeOfCalendar(mDate);
		mNotes = op.getString(op
				.getColumnIndexOrThrow(OperationTable.KEY_OP_NOTES));
		final int idx = op.getColumnIndex(OperationTable.KEY_OP_SCHEDULED_ID);
		if (idx >= 0) {
			mScheduledId = op.getLong(idx);
		} else {
			mScheduledId = 0;
		}
		final int transIdx = op
				.getColumnIndex(OperationTable.KEY_OP_TRANSFERT_ACC_ID);
		if (transIdx >= 0) {
			mTransferAccountId = op.getLong(transIdx);
			mTransSrcAccName = op.getString(op
					.getColumnIndex(OperationTable.KEY_OP_TRANSFERT_ACC_NAME));
		}

		final int accIdx = op.getColumnIndex(OperationTable.KEY_OP_ACCOUNT_ID);
		if (accIdx >= 0) {
			mAccountId = op.getLong(accIdx);
		}
	}

	public Operation() {
		mDate = new GregorianCalendar();
		Tools.clearTimeOfCalendar(mDate);
		mSum = 0L;
		mThirdParty = "";
		mMode = "";
		mTag = "";
		mNotes = "";
		mScheduledId = 0;
		mTransferAccountId = 0;
		mAccountId = 0;
	}

	public Operation(Parcel parcel) {
		mDate = new GregorianCalendar();
		Tools.clearTimeOfCalendar(mDate);
		readFromParcel(parcel);
	}

	public int getMonth() {
		return mDate.get(Calendar.MONTH);
	}

	public void setMonth(int month) {
		this.mDate.set(Calendar.MONTH, month);
	}

	public int getDay() {
		return mDate.get(Calendar.DAY_OF_MONTH);
	}

	public void setDay(int day) {
		this.mDate.set(Calendar.DAY_OF_MONTH, day);
	}

	public int getYear() {
		return mDate.get(Calendar.YEAR);
	}

	public void setYear(int year) {
		this.mDate.set(Calendar.YEAR, year);
	}

	public String getDateStr() {
		return Formater.getFullDateFormater().format(mDate.getTime());
	}

	public String getShortDateStr() {
		return Formater.getShortDateFormater(null).format(mDate.getTime());
	}

	public long getDate() {
		return mDate.getTimeInMillis();
	}

	public void setDate(long date) {
		mDate.setTimeInMillis(date);
		Tools.clearTimeOfCalendar(mDate);
	}

	public void addDay(int nbDays) {
		mDate.add(Calendar.DAY_OF_MONTH, nbDays);
	}

	public void addMonth(int nbMonths) {
		mDate.add(Calendar.MONTH, nbMonths);
	}

	public void addYear(int nbYears) {
		mDate.add(Calendar.YEAR, nbYears);
	}

	public String getSumStr() {
		return Formater.getSumFormater().format(mSum / 100.0d);
	}

	public void setSumStr(String sumStr) {
		sumStr = sumStr.replace('+', ' ').trim();
		double d;
		try {
			d = Formater.getSumFormater().parse(sumStr).doubleValue();
		} catch (ParseException e) {
			d = 0d;
		}
		mSum = Math.round(d * 100);
	}

	public void setDateStr(String dateStr) throws ParseException {
		mDate.setTime(Formater.getFullDateFormater().parse(dateStr));

	}

	@Override
	public int describeContents() {
		return 0;
	}

	@Override
	public void writeToParcel(Parcel dest, int flags) {
		dest.writeInt(getDay());
		dest.writeInt(getMonth());
		dest.writeInt(getYear());
		dest.writeString(mThirdParty);
		dest.writeString(mTag);
		dest.writeString(mMode);
		dest.writeString(mNotes);
		dest.writeLong(mSum);
		dest.writeLong(mScheduledId);
		dest.writeLong(mTransferAccountId);
	}

	protected void readFromParcel(Parcel in) {
		setDay(in.readInt());
		setMonth(in.readInt());
		setYear(in.readInt());

		mThirdParty = in.readString();
		if (null == mThirdParty) {
			mThirdParty = "";
		}

		mTag = in.readString();
		if (null == mTag) {
			mTag = "";
		}

		mMode = in.readString();
		mNotes = in.readString();
		if (null == mMode) {
			mMode = "";
		}

		mSum = in.readLong();
		mScheduledId = in.readLong();
		mTransferAccountId = in.readLong();
	}

	public static final Parcelable.Creator<Operation> CREATOR = new Parcelable.Creator<Operation>() {
		public Operation createFromParcel(Parcel in) {
			return new Operation(in);
		}

		public Operation[] newArray(int size) {
			return new Operation[size];
		}
	};

	public boolean equals(Operation op) {
		if (null == op) {
			return false;
		}
		return mDate.equals(op.mDate) && equalsButDate(op);
	}

	public boolean equalsButDate(Operation op) {
		if (null == op) {
			return false;
		}
		return mThirdParty.equals(op.mThirdParty) && mTag.equals(op.mTag)
				&& mMode.equals(op.mMode) && mNotes.equals(op.mNotes)
				&& mSum == op.mSum && mScheduledId == op.mScheduledId
				&& mTransferAccountId == op.mTransferAccountId;
	}
}

File: app/src/main/kotlin/fr/geobert/radis/data/Operation.kt
package fr.geobert.radis.data

import android.database.Cursor
import android.os.*
import fr.geobert.radis.db.*
import fr.geobert.radis.tools.*
import fr.geobert.radis.ui.adapter.CellState
import hirondelle.date4j.DateTime
import java.util.*

open class Operation() : ImplParcelable, Comparable<Operation> {
    override val parcels = hashMapOf<String, Any?>()
    var mDate: DateTime by parcels
    var mThirdParty: String by parcels
    var mTag: String by parcels
    var mMode: String by parcels
    var mNotes: String by parcels
    var mSum: Long by parcels
    var mScheduledId: Long by parcels
    var mRowId: Long by parcels
    var mTransSrcAccName: String by parcels
    var mIsChecked: Boolean by parcels

    // if these value are != 0, it is a transfert operation between 2 accounts
    // mTransferAccountId is the other account
    var mTransferAccountId: Long by parcels
    var mAccountId: Long by parcels

    // properties used only for ui
    var isSelected: Boolean = false
    var state: CellState = CellState.STATE_REGULAR_CELL


    init {
        mDate = DateTime.today(TIME_ZONE)
        mThirdParty = ""
        mTag = ""
        mMode = ""
        mNotes = ""
        mSum = 0L
        mScheduledId = 0
        mRowId = 0
        mTransSrcAccName = ""
        mIsChecked = false
        mTransferAccountId = 0L
        mAccountId = 0L
    }


    constructor(op: Operation) : this() {
        initWithOperation(op)
    }

    constructor(c: Cursor) : this() {
        initWithCursor(c)
    }

    constructor(p: Parcel) : this() {
        readFromParcel(p)
    }

    companion object {
        val CREATOR: Parcelable.Creator<Operation> = object : Parcelable.Creator<Operation> {
            override fun createFromParcel(p: Parcel): Operation {
                return Operation(p)
            }

            override fun newArray(size: Int): Array<Operation?> {
                return arrayOfNulls(size)
            }
        }
    }

    fun initWithCursor(op: Cursor) {
        mRowId = op.getLong(op.getColumnIndex(OperationTable.KEY_OP_ROWID))
        mThirdParty = op.getString(op.getColumnIndexOrThrow(InfoTables.KEY_THIRD_PARTY_NAME)) ?: ""
        mMode = op.getString(op.getColumnIndexOrThrow(InfoTables.KEY_MODE_NAME)) ?: ""
        mTag = op.getString(op.getColumnIndexOrThrow(InfoTables.KEY_TAG_NAME)) ?: ""
        mSum = op.getLong(op.getColumnIndexOrThrow(OperationTable.KEY_OP_SUM))
        setDate(op.getLong(op.getColumnIndexOrThrow(OperationTable.KEY_OP_DATE)))
        mNotes = op.getString(op.getColumnIndexOrThrow(OperationTable.KEY_OP_NOTES)) ?: ""
        val idx = op.getColumnIndex(OperationTable.KEY_OP_SCHEDULED_ID)
        if (idx >= 0) {
            mScheduledId = op.getLong(idx)
        } else {
            mScheduledId = 0
        }
        val transIdx = op.getColumnIndex(OperationTable.KEY_OP_TRANSFERT_ACC_ID)
        if (transIdx >= 0) {
            mTransferAccountId = op.getLong(transIdx)
            mTransSrcAccName = op.getString(op.getColumnIndex(OperationTable.KEY_OP_TRANSFERT_ACC_NAME)) ?: ""
        }

        val accIdx = op.getColumnIndex(OperationTable.KEY_OP_ACCOUNT_ID)
        if (accIdx >= 0) {
            mAccountId = op.getLong(accIdx)
        }
        val checkedIdx = op.getColumnIndex(OperationTable.KEY_OP_CHECKED)
        if (-1 != checkedIdx) {
            mIsChecked = op.getInt(checkedIdx) == 1
        }
    }

    fun initWithOperation(op: Operation) {
        setDate(op.getDate())
        mThirdParty = op.mThirdParty
        mTag = op.mTag
        mMode = op.mMode
        mNotes = op.mNotes
        mSum = op.mSum
        mScheduledId = op.mScheduledId
        mTransferAccountId = op.mTransferAccountId
        mTransSrcAccName = op.mTransSrcAccName
        mAccountId = op.mAccountId
        mIsChecked = op.mIsChecked
    }

    fun getMonth(): Int {
        return mDate.month
    }

    fun setMonth(month: Int) {
        val d = Math.min(mDate.day, DateTime.forDateOnly(mDate.year, month, 1).endOfMonth.day)
        mDate = DateTime.forDateOnly(mDate.year, month, d)
    }

    fun getDay(): Int {
        return mDate.day
    }

    fun setDay(day: Int) {
        val d = Math.min(day, mDate.endOfMonth.day)
        mDate = DateTime.forDateOnly(mDate.year, mDate.month, d)
    }

    fun getYear(): Int {
        return mDate.year
    }

    fun setYear(year: Int) {
        // will throw exception if month = feb and day = 29 and year is not leap, adjust the day
        val d = if (mDate.month == 2 && mDate.isLeapYear && !DateTime.forDateOnly(year, 1, 1).isLeapYear)
            Math.min(mDate.day, 28)
        else
            mDate.day
        mDate = DateTime.forDateOnly(year, mDate.month, d)
    }

    fun getDate(): Long {
        return mDate.getMilliseconds(TIME_ZONE)
    }

    fun setDate(date: Long) {
        val d = DateTime.forInstant(date, TIME_ZONE)
        mDate = DateTime.forDateOnly(d.year, d.month, d.day)
    }

    fun getDateObj(): Date {
        return Date(mDate.getMilliseconds(TIME_ZONE))
    }

    fun addDay(nbDays: Int) {
        mDate = mDate.plusDays(nbDays)
    }

    fun addMonth(nbMonths: Int) {
        mDate = mDate.plusMonth(nbMonths)
    }

    fun addYear(nbYears: Int) {
        mDate.plus(nbYears, 0, 0, 0, 0, 0, 0, DateTime.DayOverflow.LastDay)
    }

    fun getSumStr(): String {
        return (mSum.toDouble() / 100.0).formatSum()
    }

    fun setSumStr(sumStr: String) {
        mSum = sumStr.extractSumFromStr()
    }

    //    public fun setDateStr(dateStr: String) {
    //        mDate.setTime(dateStr.parseDate())
    //    }

    override fun describeContents(): Int {
        return 0
    }

    open fun equals(op: Operation?): Boolean {
        if (null == op) {
            return false
        }
        return mDate == op.mDate && equalsButDate(op)
    }

    fun equalsButDate(op: Operation?): Boolean {
        if (null == op) {
            return false
        }
        return mThirdParty == op.mThirdParty && mTag == op.mTag && mMode == op.mMode && mNotes == op.mNotes &&
                mSum == op.mSum && mScheduledId == op.mScheduledId && mTransferAccountId == op.mTransferAccountId &&
                mIsChecked == op.mIsChecked
    }

    // comparator for sorting purpose
    override fun compareTo(other: Operation): Int {
        return if (mDate.gt(other.mDate)) {
            1
        } else if (mDate.equals(other.mDate)) {
            if (mRowId >= other.mRowId) {
                1
            } else {
                -1
            }
        } else {
            -1
        }
    }

}
